crates/sui-core/src/quorum_driver/mod.rs

```diff
@@ -747,14 +747,17 @@ where
         let is_single_writer_tx = !transaction.contains_shared_object();
 
         let timer = Instant::now();
-        let tx_cert = match tx_cert {
+        let (tx_cert, newly_formed) = match tx_cert {
             None => match quorum_driver
                 .process_transaction(transaction.clone(), client_addr)
                 .await
             {
-                Ok(ProcessTransactionResult::Certified(tx_cert)) => {
+                Ok(ProcessTransactionResult::Certified {
+                    certificate,
+                    newly_formed,
```

Comment:
LGTM!

I just have a question. In the case where `newly_formed` is false, the client has already received an executed cert, is it still necessary to go through the `process_certificate` phase?

Reply:
NVM, I guess we want to make sure that that cert is executed in all validators...

Reply:
We only get back one effects from a validator. In order to get 2f+1 effects, we still need to go through process_certificate.